compiler/luci/pass/src/QuantizeWeightsWithGPTQPass.cpp

```diff
@@ -0,0 +1,680 @@
+ * @details Compensate for the quantization error and update weights using Hessian matrix
+ *
+ */
+class QuantizeWeightsWithGPTQ final : public luci::CircleNodeMutableVisitor<void>
+{
+public:
+  QuantizeWeightsWithGPTQ(
+    loco::DataType input, loco::DataType output, QuantizationGranularity granularity,
+    std::unordered_map<const luci::CircleNode *, std::vector<float>> *hessian_map)
+    : input_type(input), output_type(output), granularity(granularity), _hessian_map(hessian_map)
+  {
+  }
+
+private:
+  loco::DataType input_type;
+  loco::DataType output_type;
+  QuantizationGranularity granularity;
+  std::unordered_map<const luci::CircleNode *, std::vector<float>> *_hessian_map;
+
+  void fake_quantize_cwq(luci::CircleConst *weights, std::vector<float> &hessian) const
+  {
+    if (output_type != loco::DataType::U8)
+    {
+      throw std::runtime_error("GPTQ quantization supports u8");
+    }
+    // Find min/max per channel
+    std::vector<float> min;
+    std::vector<float> max;
+
+    cal_minmax_per_channel(weights, min, max);
+
+    std::vector<float> nudged_min(min.size());
+    std::vector<float> nudged_max(min.size());
+    std::vector<float> scaling_factor(min.size());
+    std::vector<int64_t> zp(min.size());
+
+    asymmetric_wquant_per_channel(weights, min, max, scaling_factor, zp, nudged_min, nudged_max,
+                                  output_type, hessian);
+    asymmetric_wdequant_per_channel(weights, scaling_factor, nudged_min);
+
+    auto quantparam = std::make_unique<CircleQuantParam>();
+    quantparam->min = nudged_min;
+    quantparam->max = nudged_max;
+    quantparam->scale = scaling_factor;
+    quantparam->zerop = zp;
+
+    weights->quantparam(std::move(quantparam));
+  }
+
+  void fake_quantize(luci::CircleConst *weights, std::vector<float> &hessian) const
+  {
+    switch (granularity)
+    {
+      case luci::QuantizationGranularity::ChannelWise:
+        fake_quantize_cwq(weights, hessian);
+        break;
+      default:
+        throw std::invalid_argument("Unsupported granularity");
+    }
+  }
+
+private:
+  // Check if
+  // 1. node is const
+  // 2. node's dtype is float32
+  bool is_quantizable(loco::Node *node)
+  {
+    auto const_node = dynamic_cast<luci::CircleConst *>(node);
+    if (not const_node)
+      return false;
+
+    // Skip if this is not float32
+    if (const_node->dtype() != loco::DataType::FLOAT32)
+      return false;
+
+    return true;
+  }
+
+  // Default behavior (Do nothing)
+  void visit(luci::CircleNode *) {}
+
+  void visit(luci::CircleConv2D *node)
+  {
+    LOGGER(l);
+    INFO(l) << "QuantizeWeightsWithGPTQPass visit node: " << node->name() << std::endl;
+
+    if (not is_quantizable(node->filter()))
+      return;
+
+    auto weights = loco::must_cast<luci::CircleConst *>(node->filter());
+    auto new_weights = luci::clone(weights);
+    node->filter(new_weights);
+
+    auto hessian = (*_hessian_map)[node];
+
+    fake_quantize(new_weights, hessian);
```

Comment:
looks like `weights` values are modified inside `fake_quantize`.
this is not safe if this `weights` is shared CircleConst.
plz make a clone and use it.